File: LAB#6/LAB6.5_CPP/lab6.5.cpp
#include <iostream>
#include <list>
#include <algorithm>
#include <typeinfo>
#include <string>

using namespace std;

// helper methods //
void printIntList(list<int> list) {
    string result = "{";
    for (int it : list) {
        result += to_string(it) + ", ";
    }
    // result.erase(result.size()-2);
    result += "}";
    cout<<result<<endl;
}

void printStrList(list<string> list) {
    string result = "{";
    for (string it : list) {
        result += it + ", ";
    }
    // result.erase(result.size()-2);
    result += "}";
    cout<<result<<endl;
}

string toLower(string str) {
    string result;
    for (char c : str) {
        result += std::tolower(c);
    }
    return result;
}

string reverse(string str) {
    int n = str.length();
    for (int i = 0; i < n / 2; i++) {
        swap(str[i], str[n - i - 1]);
    }
    return str;
}

// task 1 //
list<int> findCommonElements1(list<int> aList1, list<int> aList2) {
    list<int> commonElements;
    for(int i : aList1) {
        for(int j : aList2) {
            if(i == j) commonElements.push_back(i);
        }
    }
    return commonElements;
}
// or
list<int> findCommonElements2(list<int> aList1, list<int> aList2) {
    list<int> intersection;
    set_intersection(aList1.begin(), aList1.end(), aList2.begin(), aList2.end(), back_inserter(intersection));
    return intersection;
}

// task 2 //
list<string> palindrom1(list<string> strings) {
    list<string> returnList;
    for(string i : strings) {
        string word = toLower(i);
        int l = i.size();
        int indexCons = int(l/2);
        if((l%2==1 && word.substr(0, indexCons) == reverse(word.substr(indexCons+1, l)) || (l%2==0 && word.substr(0, indexCons) == reverse(word.substr(indexCons, l)))))
            returnList.push_back(i);
    }
    return returnList;
}
// or
list<string> palindrom2(list<string> strings) {
    list<string> returnList;
    for (string str : strings) {
        string strLower = toLower(str);
        string strLowerReverse = reverse(strLower);
        if(strLower == strLowerReverse) returnList.push_back(str);
    }
    return returnList;
}

int main() {
    // task 1 //
    list<int> list1 = {1,2,3,4,5,6};
    list<int> list2 = {3,4,5,7,8,9,0};
    printIntList(findCommonElements1(list1, list2));
    // or
    printIntList(findCommonElements2(list1, list2));

    // task 2 //
    list<string> strings = {"rotator", "wow", "noon", "civic", "level", "Bob", "Ross", "Tenet", "a"};
    printStrList(palindrom1(strings));
    printStrList(palindrom2(strings));
}

File: LAB#6/LAB6.5_PYTHON/lab6.5.py
# task 1 #

def findCommonElements1(aList1, aList2):
    return [i for i in aList1 if i in aList2]
# or
def findCommonElements2(aList1, aList2):
    return list(set(aList1).intersection(set(aList2)))

# task 2 #
# Any string containing just one letter is by default a palindrome.
def palindrom1(strings):
    returnList = []
    for i in strings:
        word = i.lower()
        l = len(i)
        indexCons = int(l/2)
        if((l%2==1 and word[0:indexCons] == word[l:indexCons:-1]) or (l%2==0 and word[0:indexCons] == word[l:indexCons-1:-1])):
            returnList.append(i)
    return returnList
# or #
def palindrom2(strings):
    return [str for str in strings if str.lower() == str[::-1].lower()]

# task 3 #
# The Sieve of Eratosthenes is valid for consecutive numbers from 2 to n.
def sieveOfEratosthenes(num_list):
    primes = num_list.copy()
    primes.sort()
    p = 2
    length = len(primes) - 1
    while(p * p <= length):
        for item in primes[p-1:len(primes)]:
            if item % p == 0:
                primes.remove(item)
        p += 1
    return primes

# task 4 #
def anagrams(word, word_list):
    sorted_word_list = list(map(lambda item: ''.join(set(item)).lower(), word_list))
    word = ''.join(set(word))
    anagramList = []
    for i in range(0, len(sorted_word_list)):
        if(sorted_word_list[i] == word):
            anagramList.append(word_list[i])
    return anagramList
    

def main():
    # task 1 #
    list1 = [1,2,3,4,5,6]
    list2 = [3,4,5,7,8,9,0]
    print(findCommonElements1(list1, list2))
    # or
    print(findCommonElements2(list1, list2))

    # task 2 #
    strings = ["rotator", "wow", "noon", "civic", "level", "Bob", "Ross", "Tenet", "a"]
    print(palindrom1(strings))
    # or
    print(palindrom2(strings))

    # task 3 #
    numList = [2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,24,25,26,27,28,29,30]
    print(sieveOfEratosthenes(numList))

    # task 4 #
    word_list = ["enlists", "google", "inlets", "banana"]
    word = "listen"
    print(anagrams(word, word_list))

main()